In [55]:
import pandas as pd
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV, KFold
from sklearn.metrics import accuracy_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
import warnings
warnings.filterwarnings('ignore')

# Exploratory Data Analysis(EDA):

In [2]:
data=pd.read_csv(r"C:\Users\disha\Downloads\Neural networks (1)\Neural networks\Alphabets_data.csv")
data.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


# Summary:

In [4]:
data.describe()

xbox          ybox         width       height         onpix  \
count  20000.000000  20000.000000  20000.000000  20000.00000  20000.000000   
mean       4.023550      7.035500      5.121850      5.37245      3.505850   
std        1.913212      3.304555      2.014573      2.26139      2.190458   
min        0.000000      0.000000      0.000000      0.00000      0.000000   
25%        3.000000      5.000000      4.000000      4.00000      2.000000   
50%        4.000000      7.000000      5.000000      6.00000      3.000000   
75%        5.000000      9.000000      6.000000      7.00000      5.000000   
max       15.000000     15.000000     15.000000     15.00000     15.000000   

               xbar          ybar         x2bar         y2bar         xybar  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.897600      7.500450      4.628600      5.178650      8.282050   
std        2.026035      2.325354      2.699968      2.380823      2.488475   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.000000      6.000000      3.000000      4.000000      7.000000   
50%        7.000000      7.000000      4.000000      5.000000      8.000000   
75%        8.000000      9.000000      6.000000      7.000000     10.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

            x2ybar        xy2bar         xedge        xedgey         yedge  \
count  20000.00000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.45400      7.929000      3.046100      8.338850      3.691750   
std        2.63107      2.080619      2.332541      1.546722      2.567073   
min        0.00000      0.000000      0.000000      0.000000      0.000000   
25%        5.00000      7.000000      1.000000      8.000000      2.000000   
50%        6.00000      8.000000      3.000000      8.000000      3.000000   
75%        8.00000      9.000000      4.000000      9.000000      5.000000   
max       15.00000     15.000000     15.000000     15.000000     15.000000   

            yedgex  
count  20000.00000  
mean       7.80120  
std        1.61747  
min        0.00000  
25%        7.00000  
50%        8.00000  
75%        9.00000  
max       15.00000

# Checking for Missing Values

In [48]:
data.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

# Feature and Class Information

In [5]:
num_samples = data.shape[0]
num_features = data.shape[1]

In [6]:
print(f"\nNumber of Samples: {num_samples}")
print(f"Number of Features: {num_features}")



Number of Samples: 20000
Number of Features: 17


In [41]:
classes = data['letter'].unique()
print("\nClasses:", classes)
print("Number of classes:", len(classes))


Classes: [19  8  3 13  6 18  1  0  9 12 23 14 17  5  2  7 22 11 15  4 21 24 16 20
 10 25]
Number of classes: 26


# Data Preprocessing

In [42]:
X = data.drop('letter', axis=1)
y = data['letter']

In [43]:
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

In [44]:
scaler = StandardScaler()
X = scaler.fit_transform(X)

# Creating the Model

In [45]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [49]:
def create_model(layers=(128, 64), activation='relu'):
    model = Sequential()
    model.add(Dense(layers[0], input_dim=X.shape[1], activation=activation))
    model.add(Dense(layers[1], activation=activation))
    model.add(Dense(len(classes), activation='softmax'))
    model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

In [56]:
model = KerasClassifier(build_fn=create_model, verbose=0)

# Hyperparameter Tuning

In [51]:
param_grid = {
    'layers': [(128, 64), (256, 128)],
    'activation': ['relu', 'tanh'],
    'batch_size': [32, 64],
    'epochs': [20]
}
grid = GridSearchCV(estimator=model, param_grid=param_grid, cv=3)
grid_result = grid.fit(X_train, y_train)

print("Best Accuracy:", grid_result.best_score_)
print("Best Parameters:", grid_result.best_params_)

Best Accuracy: 0.9483751455942789
Best Parameters: {'activation': 'relu', 'batch_size': 64, 'epochs': 20, 'layers': (256, 128)}


The best accuracy achieved during the grid search is approximately 94.84%.

# Evaluation:

In [52]:
y_pred = grid.predict(X_test)
y_pred_letters = label_encoder.inverse_transform(y_pred)

In [53]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')
print("\nEvaluation Metrics:")
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)


Evaluation Metrics:
Accuracy: 0.96
Precision: 0.9606746908397783
Recall: 0.96
F1 Score: 0.9599536845998847


96% of the test samples are correctly classified

The precision score of approximately 96.07% indicates that when the model predicted a particular letter, it was correct about 96.07% of the time.

The recall score of 96% indicates that the model correctly identified 96% of all the actual instances of each letter.

The F1 score of approximately 95.99% balances the precision and recall, providing a single metric for the model's accuracy.

The ANN model, after hyperparameter tuning, achieved a high accuracy of 96% on the test set.

These metrics collectively suggest that the tuned ANN model performs well in classifying the alphabet data, accurately identifying the majority of the letters in the dataset.